### ISSUE TITLE
Pre- + post-norm blocks

### ISSUE THREAD

**Issue by youngsm** [prs-eth/LitePT#20]
Hi authors,

I just wanted to say that this is really fantastic work. I had a question about part of your implementation that deviate from PTv3.

When `enable_conv=True`,  the residual CPE block is used as normal. But when it's turned off, you add a new, non-residual LayerNorm in it's place.

```python
if self.enable_conv:
    shortcut = point.feat
    point = self.conv(point)
    point.feat = shortcut + point.feat        # residual
else:
    point = self.norm0(point)                 # non-residual: x <- LN(x)
```

This makes the block roughly equivalent to a pre- + post-norm transformer (where the post-norm is just a part of the start of the next block). 

In the text you write "We observe that adding an extra LayerNorm before the attention block further stabilizes the training." However this is quite a significant change. Do you have any comparisons in terms of what you meant by "training stability"? Or how this impacts downstream performance in the end?

Cheers
Sam

**Comment by ywyue** [comment]
Hi @youngsm, thank you for your interest and for the kind words.

During development, I observed without that extra LayerNorm, sometimes the feat norm for individual encoder layers appeared very large (e.g. on the order of hundreds), which usually indicates gradient instability and occasionally lead to training crash.

To monitor or diagnose the training behavior, you can compute and log the feat norm for individual encoder layers (e.g. after this [line](https://github.com/prs-eth/LitePT/blob/079709859f02460a5ddbb9ef38f2ba01eabc9f3e/models/litept/litept.py#L405)):
```
# tracking feature norm
ms_points = []
tmp_point = point
ms_points.append(tmp_point)
for _ in range(4):
    ms_points.append(tmp_point.pooling_parent)
    tmp_point = tmp_point.pooling_parent

ms_points.reverse()

track_dict = {}
for i in range(len(ms_points)):
    track_dict["enc_{}_feat_norm".format(i)] = ms_points[i].feat.norm(p=2, dim=-1).mean().item()
```

In practice, you could also try removing the extra LayerNorm and compare the results, as whether it helps is largely empirical and may depend on your dataset and downstream task.

**Comment by youngsm** [comment]
Thanks for the reply. Yeah it would be nice to see a sort of comparison with and without this extra norm, because it really does change the training dynamics. It's interesting to hear your feature norms getting so high.

I actually ended up investigating this extra LN while diagnosing feature collapse issues I had (with SSL, where we usually use quite a bit of weight decay), where I found a few per-channel feature std's approaching zero. Empirically, removing this extra LN fixed these issues.

In the past I've also had issues with gradients/training stability. FWIW, it seems (for me and my very specific use cases) that it was mostly resolved by turning off the bf16 autocasting in CPE blocks, and also excluding all LNs/biases/1d params from participating in weight decay.

Anyways thanks for the info! Looking forward to your future work. :)

Sam

**Comment by ywyue** [comment]
Thanks for sharing those empirical findings, which will likely be helpful to the community!
